gapil/resolver/flow.go

```diff
@@ -54,12 +59,18 @@ func resolveFenceOrder(rv *resolver, f *semantic.Function, visitedFuncs visitedF
 			f.Order = semantic.Resolved | semantic.Pre | semantic.Post
 		}
 	}
+	if t.recursive && f.Order.Pre() && f.Order.Post() {
+		rv.errorf(f, "Fence in recursive function")
+		return 0
+	}
 	return f.Order
 }
 
 type fenceTracker struct {
 	resolver     *resolver
 	explicit     bool // explcitly declared fence found.
+	implicit     bool // implicitly declared fence found.
```

Comment:
Again, this confuses me. The rules on the implicit fence are done in `resolveFenceOrder` after the call to `analyse` - I'm confused how you're trying to determine the implicit fence status before we know all the facts.

Reply:
Refactored, and fixed now. Tests have also been added